In [1]:
# Для анализа данных
import numpy as np
import pandas as pd

In [3]:
df = pd.read_excel('auto_complectations.xlsx', sheet_name='Sheet1')

In [4]:
df.head()

brand     model               complect        type_car  seats  doors  \
0  Audi  A3 Sedan          1.4 TFSI 7AMT  Седан класса C      5      4   
1  Audi  A3 Sedan    sport 1.4 TFSI 7AMT  Седан класса C      5      4   
2  Audi  A3 Sedan          2.0 TFSI 7AMT  Седан класса C      5      4   
3  Audi  A3 Sedan  2.0 TFSI 7AMT quattro  Седан класса C      5      4   
4  Audi  A3 Sedan    sport 2.0 TFSI 7AMT  Седан класса C      5      4   

               Двигатель   Мощность Крутящий момент двигателя  \
0  бензиновый (1395 см³)  150 л. с.                   250 Н·м   
1  бензиновый (1395 см³)  150 л. с.                   250 Н·м   
2  бензиновый (1984 см³)  190 л. с.                   320 Н·м   
3  бензиновый (1984 см³)  190 л. с.                   320 Н·м   
4  бензиновый (1984 см³)  190 л. с.                   320 Н·м   

                 Коробка передач  ... Разгон до сотни Максимальная скорость  \
0  роботизированная (7 ступеней)  ...     8.2 секунды              224 км/ч   
1  роботизированная (7 ступеней)  ...     8.2 секунды              224 км/ч   
2  роботизированная (7 ступеней)  ...     6.8 секунды              250 км/ч   
3  роботизированная (7 ступеней)  ...     6.2 секунды              242 км/ч   
4  роботизированная (7 ступеней)  ...     6.8 секунды              242 км/ч   

  Расход топлива (л/100 км)\nгород / трасса / смешанный Дорожный просвет  \
0                                    5.9 / 4.1 / 4.8              165 мм   
1                                    5.9 / 4.1 / 4.8              165 мм   
2                                    7.2 / 4.7 / 5.6              165 мм   
3                                    7.2 / 4.8 / 5.7              165 мм   
4                                    7.2 / 4.7 / 5.6              165 мм   

  Габариты (длина × ширина × высота) Колёсная база Объём багажника  \
0                 4458 × 1796 × 1416       2637 мм           425 л   
1                 4458 × 1796 × 1416       2637 мм           425 л   
2                 4458 × 1796 × 1416       2637 мм           425 л   
3                 4458 × 1796 × 1416       2637 мм           390 л   
4                 4458 × 1796 × 1416       2637 мм           425 л   

  Объём багажника максимальный Объём топливного бака Масса автомобиля  
0                        880 л                  50 л          1320 кг  
1                        880 л                  50 л          1320 кг  
2                        880 л                  50 л          1395 кг  
3                        845 л                  55 л          1465 кг  
4                        880 л                  50 л          1395 кг  

[5 rows x 21 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1703 entries, 0 to 1702
Data columns (total 21 columns):
 #   Column                                                Non-Null Count  Dtype 
---  ------                                                --------------  ----- 
 0   brand                                                 1703 non-null   object
 1   model                                                 1703 non-null   object
 2   complect                                              1703 non-null   object
 3   type_car                                              1703 non-null   object
 4   seats                                                 1703 non-null   int64 
 5   doors                                                 1703 non-null   int64 
 6   Двигатель                                             1703 non-null   object
 7   Мощность                                              1703 non-null   object
 8   Крутящий момент двигателя                             1703 non-null 

##### 1. Название столбцов желательно привести к  единому виду. Сейчас очень неудобно

In [11]:
df.columns

Index(['brand', 'model', 'complect', 'type_car', 'seats', 'doors', 'Двигатель',
       'Мощность', 'Крутящий момент двигателя', 'Коробка передач', 'Привод',
       'Разгон до сотни', 'Максимальная скорость',
       'Расход топлива (л/100 км)\nгород / трасса / смешанный',
       'Дорожный просвет', 'Габариты (длина × ширина × высота)',
       'Колёсная база', 'Объём багажника', 'Объём багажника максимальный',
       'Объём топливного бака', 'Масса автомобиля'],
      dtype='object')

##### Создадим список англоязычный названий

In [36]:
eng_col_name = ['engine', 'power', 'engine_torque', 'transmission', 'drive_gear',
                '0_100time', 'max_speed', 'fuel_con_CHM', 'G_clearance', 'dimensions_LWH',
                'wheelbase', 'trunk_vol', 'max_trunk_vol','fuel_tank', 'vehicle_weight'
               ] # CHM в fuel_con_CHM - city / highway / mixed, LWH в dimensions_LWH - length × width × height
# Провекра
len(df.columns[6:]) == len(eng_col_name)

True

In [50]:
list_for_trans = []
for col_name in df.columns[6:]:
    list_for_trans.append(col_name)
    
dict_col_name = dict(map(lambda *args: args, list_for_trans, eng_col_name))
# Замена
df = df.rename(dict_col_name, axis='columns')

# Проверка
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1703 entries, 0 to 1702
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   brand           1703 non-null   object
 1   model           1703 non-null   object
 2   complect        1703 non-null   object
 3   type_car        1703 non-null   object
 4   seats           1703 non-null   int64 
 5   doors           1703 non-null   int64 
 6   engine          1703 non-null   object
 7   power           1703 non-null   object
 8   engine_torque   1703 non-null   object
 9   transmission    1703 non-null   object
 10  drive_gear      1703 non-null   object
 11  0_100time       1703 non-null   object
 12  max_speed       1703 non-null   object
 13  fuel_con_CHM    1703 non-null   object
 14  G_clearance     1703 non-null   object
 15  dimensions_LWH  1703 non-null   object
 16  wheelbase       1703 non-null   object
 17  trunk_vol       1703 non-null   object
 18  max_trun

##### 2. Информация о датасете

- всего 21 колонка

- 1703 строк

- числовой тип в двух колонках. Остальные текстовые, их нужно будет перевести в числовые. (Хотя некоторые алгоритмы классификации могут работать и с текстовыми значениями)

- Значения в колонках "расход топлива" и "габариты" нужно будет разбить на отдельные признаки или превратить в один, используя среднее или сумму и т.д.

##### 2.1 Признак type_car - целевая переменная.

In [55]:
# Какие значения вообще есть у этого признака
# Так-то... в большинстве случаев зная только модель автомобиля можно предсказать его тип кузова
df['type_car'].unique()

array(['Седан класса C', 'Хэтчбек класса B', 'Седан класса D',
       'Универсал повышенной проходимости класса D', 'Универсал класса D',
       'Купе класса D', 'Хэтчбек класса D', 'Седан класса E',
       'Универсал класса E', 'Хэтчбек класса E', 'Седан класса F',
       'Большой кроссовер', 'Компактный кроссовер',
       'Среднеразмерный кроссовер', 'Купе класса C', 'Спорткупе',
       'Кабриолет с мягкой крышей', 'Спорткупе класса D',
       'Спортседан класса E', 'Родстер', 'Большой внедорожник',
       'Седан класса B', 'Хэтчбек класса A', 'Кроссовер класса B', 'Вэн',
       'Горячий хэтчбек класса C', 'Внедорожник класса B',
       'Среднеразмерный внедорожник', 'Хэтчбек класса C',
       'Универсал класса C', 'Спортседан класса D', 'Купе класса E',
       'Пикап', 'Горячий седан класса C', 'Минивэн', 'Компактвэн',
       'Универсал повышенной проходимости класса E'], dtype=object)

Согласно заданию, нужно создать классификатор который будет распозновать 4 класса: 
- Седан 
- Хетчбек 
- Кроссовер 
- Внедорожник.

При этом в датасете присутсвуют и другие варианты типов кузова. Например, "универсал", "кабриолет", "купе" и т.д. Так как данных не очень много удалять эти значения не нужно.

На данный момент принято сделать следующим образом:

- Универсал, купе, кабриолет, родстер, минивен будут отнесены к единому, пятому классу "Другое"

- Пикап отнесен к классу "внедорожник"

Далее вполне можно будет отнести универсалы к хетчбекам (архитектура кузова имеет общие черты, разница в объеме багажника) 
Минивен можно отнести к классу кроссовер, а купе к седанами (однако тут нужно сразу учитыввать, что большинство купе имеют более мощный двигатель чем у обычного седана)


______________________

Сейчас оставим в колонке "type_car" только главный тип кузова без подтипов:


In [65]:
# Напишем функцию для замены названий типов кузова
def rename_type_car(type_car:str):
    '''
    Функция для замены значения в колонке "type_car"
    на класс без подклассов.
    
    На вход принимает строку с названием класса и подкласса
    Возвращает преобразованную строку.
    
    '''
    # Приведем строку к нижнему регистру 
    type_car.lower()
    
    # Назачение классов, и сразу на английском
    
    if 'седан' in type_car:
        return 'sedan'
    elif 'xетчбек' in type_car:
        return 'hatchback'
    elif 'кроссовер' in type_car:
        return 'crossover'
    elif 'внедорожник' in type_car or 'пикап' in type_car:
        return 'SUV'
    elif 'купе' in type_car:
        return 'coupe'
    elif 'вэн' in type_car:
        return 'van'
    elif 'универсал' in type_car:
        return 'wagon'
    elif 'родстер' in type_car:
        return 'roadster'
    elif 'кабриолет' in type_car:
        return 'cabriolet'
    else:
        return 'None_type'

In [9]:
# Так-то... в большинстве случаев зная только модель автомобиля можно предсказать его тип кузова, потому, что например  